src/jrd/exe.cpp

```diff
@@ -1294,18 +1295,123 @@ void EXE_execute_triggers(thread_db* tdbb,
+	const auto old_rec = old_rpb ? old_rpb->rpb_record : nullptr;
+	const auto new_rec = new_rpb ? new_rpb->rpb_record : nullptr;
+
+	if (!(dbb->dbb_flags & DBB_creating) && (old_rpb || new_rpb))
+	{
+		if (const auto relation = old_rpb ? old_rpb->rpb_relation : new_rpb->rpb_relation;
+			relation->rel_flags & REL_system)
+		{
+			if (which_trig == StmtNode::PRE_TRIG && trigger_action == TriggerAction::TRIGGER_DELETE)
+			{
+				switch ((RIDS) relation->rel_id)
+				{
+					case rel_ccon:
+						SystemTriggers::beforeDeleteCheckConstraint(tdbb, old_rec);
+						break;
+
+					case rel_indices:
+						SystemTriggers::beforeDeleteIndex(tdbb, old_rec);
+						break;
+
+					case rel_priv:
+						SystemTriggers::beforeDeleteUserPrivilege(tdbb, old_rec);
+						break;
+
+					case rel_rcon:
+						SystemTriggers::beforeDeleteRelationConstraint(tdbb, old_rec);
+						break;
+
+					case rel_rfr:
+						SystemTriggers::beforeDeleteRelationField(tdbb, old_rec);
+						break;
+
+					case rel_segments:
+						SystemTriggers::beforeDeleteIndexSegment(tdbb, old_rec);
+						break;
+
+					case rel_triggers:
+						SystemTriggers::beforeDeleteTrigger(tdbb, old_rec);
+						break;
+				}
+			}
+			else if (which_trig == StmtNode::PRE_TRIG && trigger_action == TriggerAction::TRIGGER_UPDATE)
+			{
+				switch ((RIDS) relation->rel_id)
+				{
+					case rel_ccon:
+						SystemTriggers::beforeUpdateCheckConstraint(tdbb, old_rec, new_rec);
+						break;
+
+					case rel_fields:
+						SystemTriggers::beforeUpdateField(tdbb, old_rec, new_rec);
+						break;
+
+					case rel_indices:
+						SystemTriggers::beforeUpdateIndex(tdbb, old_rec, new_rec);
+						break;
+
+					case rel_rfr:
+						SystemTriggers::beforeUpdateRelationField(tdbb, old_rec, new_rec);
+						break;
+
+					case rel_segments:
+						SystemTriggers::beforeUpdateIndexSegment(tdbb, old_rec, new_rec);
+						break;
+
+					case rel_triggers:
+						SystemTriggers::beforeUpdateTrigger(tdbb, old_rec, new_rec);
+						break;
+				}
+			}
+			else if (which_trig == StmtNode::PRE_TRIG && trigger_action == TriggerAction::TRIGGER_INSERT)
+			{
+				switch ((RIDS) relation->rel_id)
+				{
+					case rel_priv:
+						SystemTriggers::beforeInsertUserPrivilege(tdbb, new_rec);
+						break;
+
+					case rel_rcon:
+						SystemTriggers::beforeInsertRelationConstraint(tdbb, new_rec);
+						break;
+
+					case rel_refc:
+						SystemTriggers::beforeInsertRefConstraint(tdbb, new_rec);
+						break;
+				}
+			}
+			else if (which_trig == StmtNode::POST_TRIG && trigger_action == TriggerAction::TRIGGER_DELETE)
+			{
+				switch ((RIDS) relation->rel_id)
+				{
+					case rel_ccon:
+						SystemTriggers::afterDeleteCheckConstraint(tdbb, old_rec);
+						break;
+
+					case rel_rcon:
+						SystemTriggers::afterDeleteRelationConstraint(tdbb, old_rec);
+						break;
+
+					case rel_rfr:
+						SystemTriggers::afterDeleteRelationField(tdbb, old_rec);
+						break;
+				}
+			}
+		}
+	}
```

Comment:
Consider to move this block into separate routine - `SystemTriggers::execute()`, for example

Reply:
I did it with multiple execute per moment/type, letting EXE still choosing them.